FixMath2.h

```diff
@@ -0,0 +1,845 @@
+  typedef typename IntegerType<((NI+NF-1)>>3)+1>::unsigned_type internal_type ; // smallest size that fits our internal integer
+  typedef typename IntegerType<((NI+NF)>>3)+1>::unsigned_type next_greater_type ; // smallest size that fits 1<<NF for sure (NI could be equal to 0). It can be the same than internal_type in some cases.
+  
+public:
+  /** Constructor
+   */
+  UFixMath2() {;}
+
+  /** Constructor from a positive floating point value.
+@param fl Floating point value
+@return An unsigned fixed point number
+  */
+  UFixMath2(float fl)  {internal_value = /*static_cast<internal_type>*/(fl * (next_greater_type(1) << NF));}
+
+  /** Constructor from a floating point value.
+@param fl Floating point value
+@return An unsigned fixed point number
+  */
+  UFixMath2(double fl)  {internal_value = static_cast<internal_type> (fl * (next_greater_type(1) << NF)); }
+  
+  /* Constructor from integer type (as_frac = false) or from fractionnal value (as_frac=true) can be used to emulate the behavior of for instance Q8n0_to_Q8n8 */
+
+    /** Constructor from an integer value which can be interpreted as both a resulting fixed point 
+math number with a fractional part equals to 0, or as a number with decimal, ie as the underlying type behind the fixed point math number.
+@param value Integer value
+@param as_raw=false with false value will be interpreted as an integer, with true it will be interpreted as a number with decimals.
+@return An unsigned fixed point number
+  */
+  template<typename T>
+  UFixMath2(T value,bool as_raw=false)
+  {
+    if (as_raw) fromRaw(value);
+    else internal_value = (internal_type(value) << NF);
+  }
+
+
+  /** Set the internal value of the fixed point math number.
+@param raw The new internal value.
+  */  
+  template<typename T>
+  void fromRaw(T raw) { internal_value = raw; }
+
+
+
+
+
+  /** Constructor from another UFixMath2. 
+@param uf An unsigned fixed type number which value can be represented in this type.
+@return A unsigned fixed type number
+  */
+  template<byte _NI, byte _NF>
+  UFixMath2(const UFixMath2<_NI,_NF>& uf) {
+    internal_value = SHIFTR((typename IntegerType<((MAX(NI+NF-1,_NI+_NF-1))>>3)+1>::unsigned_type) uf.asRaw(),(_NF-NF));
+  }
+
+  /** Constructor from another SFixMath2. 
+@param uf An signed fixed type number which value can be represented in this type.
+@return A unsigned fixed type number
+  */
+  template<byte _NI, byte _NF>
+  UFixMath2(const SFixMath2<_NI,_NF>& uf) {
+    internal_value = SHIFTR((typename IntegerType<((MAX(NI+NF-1,_NI+_NF-1))>>3)+1>::unsigned_type) uf.asRaw(),(_NF-NF));
+  }
+
+
+  //////// ADDITION OVERLOADS
+
+  /** Addition with another UFixMath2. Safe.
+@param op The UFixMath2 to be added.
+@return The result of the addition as a UFixMath2.
+  */
+  template<byte _NI, byte _NF>
+  UFixMath2<MAX(NI,_NI)+1,MAX(NF,_NF)> operator+ (const UFixMath2<_NI,_NF>& op) const
+  {
+    constexpr byte new_NI = MAX(NI, _NI) + 1;
```

Comment:
I was a bit hesitant about the +1 at first. int addition does not do that, after all. But I guess it makes sense, esp. in combination with the earlier comment that NI does not need to be divisible by 8.

Reply:
Yes, basically this enforces that no overflow can happen, like the behavior of a float (well, to some extend). It really behaves differently from standard integer types but I think that, wisely used and well introduced it can really be neat to not care too much about overflows once things are created.